js/compiler/compiler.go

```diff
@@ -44,103 +44,114 @@ var (
+	CompatibilityModeExtended CompatibilityMode = iota + 1
+	// CompatibilityModeBase is standard goja ES5.1+
+	CompatibilityModeBase
+)
 
-// Constructs a new compiler.
-func New() (*Compiler, error) {
-	var err error
-	once.Do(func() {
-		compilerInstance, err = new()
-	})
+// A Compiler compiles JavaScript source code (ES5.1 or ES6) into a goja.Program
+type Compiler struct{}
 
-	return compilerInstance, err
+// New returns a new Compiler
+func New() *Compiler {
+	return &Compiler{}
 }
 
-func new() (*Compiler, error) {
-	conf := rice.Config{
-		LocateOrder: []rice.LocateMethod{rice.LocateEmbedded},
+// Transform the given code into ES5, while synchronizing to ensure only a single
+// babel instance is in use.
+func (c *Compiler) Transform(src, filename string) (code string, srcmap SourceMap, err error) {
+	var b *babel
+	if b, err = newBabel(); err != nil {
+		return
 	}
 
-	babelSrc := conf.MustFindBox("lib").MustString("babel.min.js")
+	b.mutex.Lock()
+	defer b.mutex.Unlock()
+	return b.Transform(src, filename)
+}
 
-	c := &Compiler{vm: goja.New()}
-	if _, err := c.vm.RunString(babelSrc); err != nil {
-		return nil, err
+// Compile the program in the given CompatibilityMode, optionally running pre and post code.
+func (c *Compiler) Compile(src, filename string, pre, post string,
+	strict bool, compatMode CompatibilityMode) (*goja.Program, string, error) {
+	code := pre + src + post
+	ast, err := parser.ParseFile(nil, filename, code, 0)
+	if err != nil {
+		if compatMode == CompatibilityModeExtended {
+			code, _, err = c.Transform(src, filename)
+			if err != nil {
+				return nil, code, err
+			}
+			return c.Compile(code, filename, pre, post, strict, compatMode)
+		}
+		return nil, src, err
 	}
+	pgm, err := goja.CompileAST(ast, strict)
+	return pgm, code, err
+}
 
-	c.this = c.vm.Get("Babel")
-	thisObj := c.this.ToObject(c.vm)
-	if err := c.vm.ExportTo(thisObj.Get("transform"), &c.transform); err != nil {
-		return nil, err
-	}
+type babel struct {
+	vm        *goja.Runtime
+	this      goja.Value
+	transform goja.Callable
+	mutex     sync.Mutex //TODO: cache goja.CompileAST() in an init() function?
+}
+
+func newBabel() (*babel, error) {
+	var err error
 
-	return c, nil
+	once.Do(func() {
+		conf := rice.Config{
+			LocateOrder: []rice.LocateMethod{rice.LocateEmbedded},
+		}
+		babelSrc := conf.MustFindBox("lib").MustString("babel.min.js")
+		vm := goja.New()
+		if _, err = vm.RunString(babelSrc); err != nil {
+			return
+		}
+
+		this := vm.Get("Babel")
+		bObj := this.ToObject(vm)
+		babl = &babel{vm: vm, this: this}
+		if err = vm.ExportTo(bObj.Get("transform"), &babl.transform); err != nil {
+			return
+		}
+	})
+
+	return babl, err
 }
 
-// Transform the given code into ES5.
-func (c *Compiler) Transform(src, filename string) (code string, srcmap SourceMap, err error) {
-	opts := make(map[string]interface{})
-	for k, v := range DefaultOpts {
-		opts[k] = v
-	}
```

Comment:
> I would prefer if we still copy it instead of changing it.

Oof, rookie mistake on my part there, nice catch. I did a quick `opts := DefaultOpts` hoping to avoid the loop, but that only copies the `map` pointer, which is not what I meant, and you're right, the values should be copied.

Will fix.

Reply:
Fixed in 9875b789.

Reply:
> To be honest there should probably be a function that returns you the map and for there to be no global whatsoever

I kind of agree, but I initially used that approach for `getStaticResolvers()` in #1143, and if I interpreted [@na-- 's comment](https://github.com/loadimpact/k6/pull/1143#discussion_r337355756) correctly, he suggested to leave it as a presumably global `map`, which I ultimately did and you approved. :)

I'm kind of torn on globals in Go, or in general I guess, since if used "carefully" they shouldn't cause any issues, and the Go stdlib is certainly a testament to that.

In cases like this, `DefaultOpts` should really be a constant or immutable object, but since there's no such thing or would require custom handling, you can either `nolint` it or use a function. I flip-flop between both, but since changing this would be more work, I'm opting for leaving it as is. Unless you convince me otherwise :)